lib/Fun_Pred_Syntax.thy

```diff
@@ -0,0 +1,187 @@
+  "pred_bot2 \<equiv> \<lambda>_ _. False"
+
+(* Syntax bundles *)
+bundle fun_pred_syntax
+begin
+  (* infixl instead of infixr, because we want to split off conjuncts from the left *)
+  notation pred_conj (infixl "and" 35)
+  notation pred_disj (infixl "or" 30)
+  notation pred_neg ("not _" [40] 40)
+  notation pred_top ("\<top>")
+  notation pred_bot ("\<bottom>")
+  notation pred_top2 ("\<top>\<top>")
+  notation pred_bot2 ("\<bottom>\<bottom>")
+end
+
+bundle no_fun_pred_syntax
+begin
+  no_notation pred_conj (infixl "and" 35)
+  no_notation pred_disj (infixl "or" 30)
+  no_notation pred_neg ("not _" [40] 40)
+  no_notation pred_top ("\<top>")
+  no_notation pred_bot ("\<bottom>")
+  no_notation pred_top2 ("\<top>\<top>")
+  no_notation pred_bot2 ("\<bottom>\<bottom>")
+end
+
+unbundle fun_pred_syntax
+
+
+(* Definition lemmas specialised to "bool" and fun instance of boolean_algebra: *)
+lemmas pred_conj_def =
+  inf_fun_def[where 'b=bool, simplified]
+  inf_fun_def[where f="f::'a \<Rightarrow> 'b::boolean_algebra" for f]
+
+lemmas pred_disj_def =
+  sup_fun_def[where 'b=bool, simplified]
+  sup_fun_def[where f="f::'a \<Rightarrow> 'b::boolean_algebra" for f]
+
+lemmas pred_neg_def =
+  fun_Compl_def[where 'b=bool, simplified]
+  fun_Compl_def[where A="A::'a \<Rightarrow> 'b::boolean_algebra" for A]
+
+lemmas pred_top_def[simp] =
+  top_fun_def[where 'b=bool, simplified] top_fun_def[where 'b="'b::boolean_algebra"]
+
+lemmas pred_bot_def[simp] =
+  bot_fun_def[where 'b=bool, simplified] bot_fun_def[where 'b="'b::boolean_algebra"]
+
+(* AC rewriting renamed and specialised, so we don't have to remember inf/sup *)
+lemmas pred_conj_aci = inf_aci[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+lemmas pred_disj_aci = sup_aci[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+
+(* Additional legacy names *)
+lemmas pred_conjI = inf1I inf2I
+
+lemmas pred_disjI1 = sup1I1[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+lemmas pred_disjI2 = sup1I2[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+
+lemmas pred_disj1CI[intro!] = sup1CI[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+lemmas pred_disj2CI = sup2CI[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+
+lemmas pred_conj_assoc = inf.assoc[where 'a="'a \<Rightarrow> 'b::boolean_algebra", symmetric]
+lemmas pred_conj_comm = inf.commute[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+
+lemmas pred_disj_assoc = sup.assoc[where 'a="'a \<Rightarrow> 'b::boolean_algebra", symmetric]
+lemmas pred_disj_comm = sup.commute[where 'a="'a \<Rightarrow> 'b::boolean_algebra"]
+
+(* Top/bot and function composition: *)
+lemma pred_top_comp[simp]:
+  "\<top> \<circ> f = \<top>"
+  by (simp add: comp_def)
+
+lemma pred_bot_comp[simp]:
+  "\<bottom> \<circ> f = \<bottom>"
+  by (simp add: comp_def)
+
+(* We would get these for free if we could instantiate pred_top/pred_bot to top/bot directly: *)
+lemmas pred_top_left_neutral[simp] =
```

Comment:
trivial: recommend newline above this line

Reply:
actually, looking further above, it seems like I'd recommend it for most of the comments that outline sections of lemmas... have you considered `text` or actual sections of some kind?

Reply:
actual (sub)sections might make sense, yes, I'll have a look at that